### DISCUSSION TITLE
Resolve CapabilityServerSet's promise

### DISCUSSION THREAD

**Discussion by iso-morph** [capnproto/capnproto#2173] [Q&A] [answered]
Hello

For an API wrapper and extension, I need to resolve a client capability back into the server object. As suggested [here](https://groups.google.com/g/capnproto/c/Iti7gWjgI8g) I use a `CapabilityServerSet`. However, `getLocalServer` returns a promise and I am not sure how to resolve this immediately such that I can access the server object's methods.

More specifically, I have an `ExtHandle` object holding some context data for later API calls. This object which is created by the client through `newExtHandle`. Later, it is passed as an argument to `doSomeWork`. On the server side, I need access to the context data by means of a pointer.

Here is a what I have, stripped to reduced overhead. The full example can be provided.

**demo.capnp**
```
interface Demo {
    interface ExtHandle {
        init @0 ();
        destroy @1 ();
    }
    newExtHandle @0 () -> (handle :ExtHandle);

    doSomeWork @1 (task :Text, handle :ExtHandle = null) -> (result :Int32);
}
```

**external_api.h**
```
typedef struct {
    bool is_initialised;
    uint32_t value;
    // more context data
} ext_handle_t;

uint32_t do_some_work(const char* task, ext_handle_t *handle);
```

**demo_server.c++**
```++
class ExtHandleImpl : public Demo::ExtHandle::Server {
        ext_handle_t ext_handle = {.is_initialised = false, .value = 0};

    public:
        ext_handle_t* get_handle_pointer(void) {
            return &(this->ext_handle);
        }
}

class DemoServerImpl : public Demo::Server {
        capnp::CapabilityServerSet<Demo::ExtHandle> ext_handles;
    public:
        ::kj::Promise<void> newExtHandle(NewExtHandleContext context) override {
            Demo::ExtHandle::Client handle = this->ext_handles.add(kj::heap<ExtHandleImpl>(this, thisCap()));
            context.getResults().setHandle(handle);
            return kj::READY_NOW;
        }

        ::kj::Promise<void> doSomeWork(DoSomeWorkContext context) override {
            auto task = context.getParams().getTask();
            auto hdl = context.getParams().getHandle();

            // how to resolve this promise immediately?
            kj::Maybe<ExtHandleImpl&> maybe_ext_handle = this->ext_handles.getLocalServer(hdl).wait();
            ext_handle_t *ext_handle = nullptr;
            KJ_IF_MAYBE(hdl, maybe_ext_handle) {
                ext_handle = hdl->get_handle_pointer();
            }
            uint32_t result = do_some_work(task.cStr(), ext_handle);
            context.getResults().setResult(result);
            return kj::READY_NOW;
        }
}
```

Is this even possible and is my approach correct?

Thank you very much in advance.

**Comment by kentonv** [comment]
You cannot "resolve the promise immediately", you have to wait for it, either using `.then()` or (if your compiler supports it) `co_await`.

Here's some documentation on how to do asynchronous programming in KJ:

https://github.com/capnproto/capnproto/blob/v2/kjdoc/tour.md#asynchronous-event-loop

**Reply by iso-morph** [reply]
Hi Kenton, thank you very much for your quick reply.

If I understand correctly, I cannot `.wait()` for the promise as I am already in a running event loop, namely the server's. On the other hand, I finally need to set my external API's result as the handler's result. But I am having trouble doing this in `.then()`'s lambda as I cannot pass the context to the lambda function.

Anyway, I have a feeling that I might be completely on the wrong track. The bottomline is that I need to resolve a `Demo::ExtHandle::Client` capability, which was passed to the client and back, back into the corresponding server object `ExtHandleImpl()` in order to access the handle context data it holds.

Perhaps there is a better way of achieving this? Given the fact that the object was created on the server and is thus known to exist. Any advice is greatly appreciated.

**Reply by iso-morph** [reply, marked as answer]
I figured it out! The "trick" is to return the promise instead of `kj::READY_NOW`. Thank you Kenton for your active support on various platforms. It was [this](https://stackoverflow.com/questions/73681340/capnproto-request-to-other-server-in-callback#comment130127412_73685529) comment of yours which finally lead me to the right track.

For future reference, I include my implementation of `doSomeWork`.

```c++
        ::kj::Promise<void> doSomeWork(DoSomeWorkContext context) override {
            auto task = context.getParams().getTask();
            auto hdl = context.getParams().getHandle();

            return this->ext_handles.getLocalServer(hdl).then(
                [](kj::Maybe<Demo::ExtHandle::Server&> ext_handle) {
                    KJ_IF_MAYBE(hdl, ext_handle) {
                        return static_cast<ExtHandleImpl*>(hdl);
                    } else {
                        KJ_FAIL_ASSERT("cannot resolve ExtHandle");
                    }
                }
            ).then(
                [](ExtHandleImpl* ext_handle) {
                    return ext_handle->get_handle_pointer();
                }
            ).then(
                [task](ext_handle_t* hdl) {
                    return do_some_work(task.cStr(), hdl);
                }
            ).then(
                [KJ_CPCAP(context)](uint32_t result) mutable {
                    context.getResults().setResult(result);
                    return;
                }
            );
```

I mark this as the answer as it includes the solution. But credits are due to Kenton!
